homeassistant/components/device_tracker/fortinet_fortios.py

```diff
@@ -0,0 +1,220 @@
+"""
+Support for Fortigate/FortiWifi Routers.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/device_tracker.fortios/
+"""
+import logging
+
+import voluptuous as vol
+
+import homeassistant.helpers.config_validation as cv
+from homeassistant.components.device_tracker import (
+    DOMAIN, PLATFORM_SCHEMA, DeviceScanner)
+from homeassistant.const import CONF_HOST, CONF_PASSWORD, CONF_USERNAME, \
+    CONF_PORT
+from homeassistant.util import slugify
+
+_LOGGER = logging.getLogger(__name__)
+
+CONF_VDOM = 'vdom'
+
+REQUIREMENTS = ['pexpect==4.0.1']
+
+PLATFORM_SCHEMA = vol.All(
+    PLATFORM_SCHEMA.extend({
+        vol.Required(CONF_HOST): cv.string,
+        vol.Required(CONF_USERNAME): cv.string,
+        vol.Optional(CONF_PASSWORD, default=''): cv.string,
+        vol.Optional(CONF_PORT): cv.port,
+        vol.Optional(CONF_VDOM): cv.string,
+    })
+)
+
+DEFAULT_DEVICE_IDLE_TIMEOUT = 300
+
+
+def get_scanner(hass, config):
+    """Validate the configuration and return a Fortinet scanner."""
+    scanner = FortinetDeviceScanner(config[DOMAIN])
+
+    return scanner if scanner.success_init else None
+
+
+class FortinetDeviceScanner(DeviceScanner):
+    """This class queries a router running Fortinet fortios firmware."""
+
+    def __init__(self, config):
+        """Initialize the scanner."""
+        self.host = config.get(CONF_HOST)
+        self.username = config.get(CONF_USERNAME)
+        self.port = config.get(CONF_PORT)
+        self.password = config.get(CONF_PASSWORD)
+        self.vdom = config.get(CONF_VDOM)
+
+        self.last_results = {}
+
+        self.success_init = self._update_info()
+        _LOGGER.info('fortinet_fortios scanner initialized')
+
+    # pylint: disable=no-self-use
+    def get_device_name(self, device):
+        """Get the firmware doesn't save the name of the wireless device."""
+        return None
+
+    def scan_devices(self):
+        """Scan for new devices and return a list with found device IDs."""
+        self._update_info()
+
+        return self.last_results
+
+    def _update_info(self):
```

Comment:
There is a lot of FortiOS specific code. This code should be moved to it's own module which is [published](https://medium.freecodecamp.org/how-to-publish-a-pyton-package-on-pypi-a89e9522ce24) on PyPI and imported here.